BinaryEncoding.md

```diff
@@ -317,18 +317,20 @@ It is legal to have several entries with the same type.
 | Name | Opcode | Immediate | Description |
 | ---- | ---- | ---- | ---- |
 | `nop` | `0x00` | | no operation |
-| `block` | `0x01` | count = `varuint32` | a sequence of expressions, the last of which yields a value |
-| `loop` | `0x02` | count = `varuint32` | a block which can also form control flow loops |
-| `if` | `0x03` | | high-level one-armed if |
-| `if_else` | `0x04` | | high-level two-armed if |
+| `block` | `0x01` |  | begin a sequence of expressions, the last of which yields a value |
+| `loop` | `0x02` |  | begin a block which can also form control flow loops |
+| `if` | `0x03` | | begin if expression |
+| `else` | `0x04` | | begin else expression of if |
 | `select` | `0x05` | | select one of two values based on condition |
 | `br` | `0x06` | relative_depth = `varuint32` | break that targets a outer nested block |
 | `br_if` | `0x07` | relative_depth = `varuint32` | conditional break that targets a outer nested block |
 | `br_table` | `0x08` | see below | branch table control flow construct |
-| `return` | `0x14` | | return zero or one value from this function |
-| `unreachable` | `0x15` | | trap immediately |
+| `return` | `0x09` | | return zero or one value from this function |
+| `unreachable` | `0x0a` | | trap immediately |
+| `end` | `0x0f` | | end a block, loop, or if |
```

Comment:
@kripken It's possible we a talking about different aspects of the 'stack' because the arity of the break operators seems to have nothing to do with the block AST encoding, so it is not clear where this '0/1' issue is entering the discussion. In your example at the end of the block an AST node representing the block would be pushed on the node stack so it would be `A B C (block D E F)`. Consider the following example of an encoding and its decoding:

```
(block $l1
  (i32.const 1)
  (i32.const 2)
  (block $l2
    (i32.const 3)
    (i32.const 4)
    (if (i32.const 5)
	(then (br $l2))
	(else (br $l2 (i32.const 6))))
    (call $fn_returning_two_values)
    (nop))
  (nop))

block_start $l1
  i32.const 1      Stack $l1: (i32.const 1)
  i32.const 2      Stack $l1: (i32.const 1) (i32.const 2)
  block_start $l2
    i32.cont 3     Stack $l1: (i32.const 1) (i32.const 2) Stack $l2: (i32.const 3)
    i32.cont 4     ... Stack $l2: (i32.const 3) (i32.const 4)
    i32.cont 5     ... Stack $l2: (i32.const 3) (i32.const 4) (i32.const 5)
    if_start       ... Stack $l2: (i32.const 3) (i32.const 4) (if (i32.const 5) (then) (else)) Stack then:
      br $l2       ... Stack then: (br $l2 (nop))
    else           ... Stack $l2: (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else)) Stack else:
      i32.const 6  ... Stack else: (i32.const 6)
    end_if         ... Stack $l2: (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6)))
    call $fn_returning_two_value  ... Stack $l2: (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6))) (call $fn_returning_two_values)
    nop            ... Stack $l2: (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6))) (call $fn_returning_two_values) (nop)
  block_end        Stack $l1 (i32.const 1) (i32.const 2) (block $l2 (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6))) (call $fn_returning_two_values) (nop))
  nop              Stack $l1 (i32.const 1) (i32.const 2) (block $l2 (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6))) (call $fn_returning_two_values) (nop)) (nop)
block_end          result: (block $l1 (i32.const 1) (i32.const 2) (block $l2 (i32.const 3) (i32.const 4) (if (i32.const 5) (then (br $l2 (nop))) (else (i32.const 6))) (call $fn_returning_two_values) (nop)) (nop))
```

The arity of the break operators do not seem to affect the decoding, nor did a call returning multiple values. AST nodes and expressions are always represented by a single AST node irrespective of the number of values the expression returns. If I have missed some concern then an example might help?